pypit/wavetilts.py

```diff
@@ -0,0 +1,480 @@
+             ii.  Fit the individual arc lines
+             iii.  2D Fit to the offset from pixcen
+             iv. Save
+
+        Parameters
+        ----------
+        maskslits : ndarray (bool), optional
+        doqa : bool
+        wv_calib : dict
+
+        Returns
+        -------
+        self.final_tilts
+        maskslits
+        """
+        # If the user sets no tilts, return here
+        if self.settings['tilts']['method'].lower() == "zero":
+            # Assuming there is no spectral tilt
+            self.final_tilts = np.outer(np.linspace(0.0, 1.0, self.msarc.shape[0]), np.ones(self.msarc.shape[1]))
+            return self.final_tilts, None, None
+
+        if maskslits is None:
+            maskslits = np.zeros(self.nslit, dtype=bool)
+
+        # Extract the arc spectra for all slits
+        self.arccen, self.arc_maskslit = self._extract_arcs()
+
+        # maskslit
+        self.mask = maskslits & (self.arc_maskslit==1)
+        gdslits = np.where(self.mask == 0)[0]
+
+        # Final tilts image
+        self.final_tilts = np.zeros_like(self.msarc)
+        # Loop on all slits
+        for slit in gdslits:
+            # Trace
+            _ = self._trace_tilts(slit, wv_calib=wv_calib)
+
+            # Model line-by-line
+            _ = self._analyze_lines(slit)
+
+            # 2D model of the tilts
+            #   Includes QA
+            self.tilts = self._fit_tilts(slit, doqa=doqa)
+
+            # Save to final image
+            word = self.slitpix == slit+1
+            self.final_tilts[word] = self.tilts[word]
+
+        return self.final_tilts, maskslits
+
+    def _qa(self, slit):
+        """
+        QA
+          Wrapper to artraceslits.slit_trace_qa()
+
+        Parameters
+        ----------
+        slit : int
+
+        Returns
+        -------
+
+        """
+        self.tiltsplot, self.ztilto, self.xdat = artracewave.prep_tilts_qa(
+            self.msarc, self.all_ttilts[slit], self.tilts, self.all_trcdict[slit]['arcdet'],
+            self.pixcen, slit)
+
+    def save_master(self, outfile=None):
+        """
+
+        Parameters
+        ----------
+        outfile
+        use_tilts_as_final
+
+        Returns
+        -------
+
+        """
+        if outfile is None:
+            outfile = self.ms_name
+        #
+        if self.final_tilts is None:
+            msgs.warn("final_tilts not yet created.  Make it!")
+            return
+        #
+        hdu0 = fits.PrimaryHDU(self.final_tilts)
+        hdul = [hdu0]
+
+        for slit in range(self.nslit):
+            # Bad slit?
+            if self.mask[slit]:
+                continue
+            # fweight and model
+            xtfits = self.all_trcdict[slit]['xtfit']  # For convenience
+            xszs = [len(xtfit) if xtfit is not None else 0 for xtfit in xtfits]
+            maxx = np.max(xszs)
+            # Add 1 to pack in ycen
+            fwm_img = np.zeros((maxx+1, len(xtfits), 4)) - 9999999.9
```

Comment:
I feel like `- 9999999.9` should probably be defined in the function definition as a variable called `maskValue` or something similar.

Eventually, these things will probably become masked arrays.